File: src/stages/link.js
import { basename, dirname, extname } from "path";
import { progressBar } from "../utils/progress.js";

// Regex for Google Takeout metadata suffixes (longest to shortest for greedy matching).
const metadataSuffixPattern = new RegExp(
  [
    "\\.supplemental-metadata",
    "\\.supplemental-metadat",
    "\\.supplemental-metada",
    "\\.supplemental-metad",
    "\\.supplemental-meta",
    "\\.supplemental-met",
    "\\.supplemental-me",
    "\\.supplemental-m",
    "\\.supplemental-",
    "\\.supplemental",
    "\\.supplementa",
    "\\.supplement",
    "\\.supplemen",
    "\\.suppleme",
    "\\.supplem",
    "\\.supple",
    "\\.suppl",
    "\\.supp",
    "\\.sup",
    "\\.su",
    "\\.s",
  ].join("|") + "$"
);

// Regex for -edited variations (longest to shortest for greedy matching).
const editedSuffixPattern = new RegExp(
  ["-edited", "-edite", "-edit", "-edi", "-ed", "-e"].join("|") + "$"
);

/**
 * Matches media files to their JSON sidecar metadata files.
 *
 * This stage uses a two-pass matching strategy:
 *  - Pass 1 (Exact): Matches files with 100% name + index confidence.
 *    Allows multiple media files to share a JSON (e.g. IMG.HEIC and IMG.MOV).
 *  - Pass 2 (Fuzzy): Attempts to match remaining media to remaining JSONs.
 *    Checks for truncation in both directions. Strictly prevents stealing
 *    JSONs already claimed by other fuzzy matches.
 *
 * @param {{
 *   filesMedia: string[],
 *   filesMetadata: string[],
 *   filesMetadataAlbums: string[]
 * }} rawCollections The raw file collections from the scan stage.
 * @param {{ fuzzyThreshold: number }} options Configuration for fuzzy matching.
 * @returns {{
 *   manifest: Array<{ mediaPath: string, jsonPath?: string, source: object }>,
 *   stats: { unmatchedMetadataFiles: Set<string>, unmatchedMediaFiles: Set<string> }
 * }} The manifest of linked files and matching statistics.
 */
export function link(rawCollections, options) {
  const { filesMedia, filesMetadata, filesMetadataAlbums } = rawCollections;
  const { fuzzyThreshold } = options;

  // Stage 1: index metadata by directory for O(1) lookups.
  const metadataByDir = new Map();
  for (const filePath of filesMetadata) {
    const dir = dirname(filePath);
    if (!metadataByDir.has(dir)) metadataByDir.set(dir, []);
    metadataByDir.get(dir).push({ path: filePath, ...sidecar(filePath) });
  }

  // Stage 2: index albums by directory.
  const albums = new Map();
  for (const path of filesMetadataAlbums) {
    const dir = dirname(path);
    albums.set(dir, basename(dir));
  }

  // Stage 3: pre-parse media files for performance.
  const parsedMedia = filesMedia.map((path) => ({
    originalPath: path,
    ...media(path),
  }));

  // Stage 4: two-pass matching algorithm.
  const manifest = [];
  const matchedMediaIndices = new Set();

  // Consumption tracking:
  // - usedByExact: JSONs claimed by Pass 1. Can be shared by other exact matches.
  // - usedByFuzzy: JSONs claimed by Pass 2. Exclusive (cannot be shared).
  const usedByExact = new Set();
  const usedByFuzzy = new Set();

  const bar = progressBar("⧗ Linking files |{bar}| {percentage}% | {value}/{total} files");
  bar.start(filesMedia.length, 0);

  // Stage 4.1: exact matches.
  // We match files that we are 100% sure about.
  // We allow sharing here (e.g. IMG.MP and IMG.MP.jpg share IMG.MP.jpg.json).
  for (let i = 0; i < parsedMedia.length; i++) {
    const media = parsedMedia[i];
    const match = findMatch(media, metadataByDir, usedByFuzzy, fuzzyThreshold, false);

    if (match) {
      matchedMediaIndices.add(i);
      usedByExact.add(match.path);
      manifest.push(createEntry(media.originalPath, match.path, albums));
      bar.increment();
    }
  }

  // Stage 4.2: fuzzy matches.
  // We match remaining files using fuzzy logic for truncated filenames.
  for (let i = 0; i < parsedMedia.length; i++) {
    if (matchedMediaIndices.has(i)) continue;

    const media = parsedMedia[i];
    const match = findMatch(media, metadataByDir, usedByFuzzy, fuzzyThreshold, true);

    if (match) {
      usedByFuzzy.add(match.path);
      manifest.push(createEntry(media.originalPath, match.path, albums));
    } else {
      manifest.push(createEntry(media.originalPath, undefined, albums));
    }
    bar.increment();
  }

  bar.stop();

  // Stage 5: compute statistics.
  const allUsed = new Set([...usedByExact, ...usedByFuzzy]);
  const unmatchedMetadataFiles = new Set(filesMetadata.filter((f) => !allUsed.has(f)));
  const unmatchedMediaFiles = new Set(manifest.filter((m) => !m.jsonPath).map((m) => m.mediaPath));

  return {
    manifest,
    stats: { unmatchedMetadataFiles, unmatchedMediaFiles },
  };
}

// Creates a manifest entry for a media file.
function createEntry(mediaPath, jsonPath, albums) {
  const dir = dirname(mediaPath);
  return {
    mediaPath,
    source: albums.has(dir) ? { type: "album", name: albums.get(dir) } : { type: "loose" },
    jsonPath,
  };
}

// Finds the best metadata match for a media file within the same directory.
function findMatch(media, metadataByDir, usedByFuzzy, fuzzyThreshold, allowFuzzy) {
  const dir = dirname(media.originalPath);
  const candidates = metadataByDir.get(dir) || [];

  let bestMatch = null;
  let bestScore = 0;

  for (const meta of candidates) {
    // Safety: if we are in fuzzy mode, we cannot touch JSONs that are
    // already claimed by another fuzzy match. However, we CAN share
    // sidecars claimed by exact matches (Live Photo scenario where
    // one component is truncated but should still share the sidecar).
    if (allowFuzzy && usedByFuzzy.has(meta.path)) continue;

    // Strict duplicate check: (1) must always match (1).
    // We never fuzzy match across indices.
    if (media.duplicate !== meta.duplicate) continue;

    const score = matchScore(media, meta, allowFuzzy, fuzzyThreshold);

    // In strict mode, ignore anything less than a perfect match.
    if (!allowFuzzy && score < 100) continue;

    if (score > bestScore) {
      bestScore = score;
      bestMatch = meta;
    }
  }

  return bestMatch;
}

// Calculates a match score between a media file and a metadata file.
// Score hierarchy:
//  - 150: Exact name + exact extension match (safest)
//  - 100: Exact name match OR embedded extension match (safe)
//  - 50+: Fuzzy/truncated name match (lowest priority)
//  - 0: No match
function matchScore(media, meta, allowFuzzy, fuzzyThreshold) {
  const { name: mediaName, extension: mediaExt } = media;
  const { name: metaName, extension: metaExt } = meta;

  // Case 1: direct name match.
  if (mediaName === metaName) {
    return metaExt && metaExt === mediaExt ? 150 : 100;
  }

  // Case 2: embedded extension match.
  // Handles cases where metadata includes the media extension in its name.
  // e.g. Media: "123.MP" (Name: "123", Ext: ".MP") vs Meta: "123.MP"
  const compositeName = `${mediaName}${mediaExt}`.toLowerCase();
  const normalizedMeta = metaName.toLowerCase();

  if (compositeName === normalizedMeta) {
    return 100; // Treated as an exact match.
  }

  if (!allowFuzzy) return 0;

  // Case 3: fuzzy matching requires sufficient filename length.
  // Google only truncates names > ~40 chars, so shorter names cannot be
  // truncation candidates. This prevents false positives like IMG_1234
  // matching IMG_123.json.
  const longerName = Math.max(mediaName.length, metaName.length);
  if (longerName < fuzzyThreshold) return 0;

  // Case 4: bidirectional fuzzy match.
  // Checks if A starts with B OR B starts with A.
  // Handles truncation on either file system side or Google side.
  const mediaStarts = mediaName.startsWith(metaName);
  const metaStarts = metaName.startsWith(mediaName);

  if (mediaStarts || metaStarts) {
    // Score is 50 + length of the shorter string (common prefix).
    return 50 + Math.min(mediaName.length, metaName.length);
  }

  return 0;
}

// Parses a media file path into its component parts for matching.
function media(filePath) {
  const extension = extname(filePath);
  let name = basename(filePath, extension);

  // Strip duplicate marker (N) from the end.
  const duplicateMatch = name.match(/\((\d+)\)$/);
  const duplicate = duplicateMatch ? parseInt(duplicateMatch[1]) : null;

  if (duplicateMatch) {
    name = name.slice(0, duplicateMatch.index);
  }

  // Strip -edited variants from the end.
  const editedMatch = name.match(editedSuffixPattern);
  if (editedMatch) {
    name = name.slice(0, editedMatch.index);
  }

  return { name, duplicate, extension: extension?.toLowerCase() };
}

// Parses a sidecar JSON file path into its component parts for matching.
function sidecar(filePath) {
  let name = basename(filePath, extname(filePath));

  // Strip duplicate marker (N) from the end.
  const duplicateMatch = name.match(/\((\d+)\)$/);
  const duplicate = duplicateMatch ? parseInt(duplicateMatch[1]) : null;

  if (duplicateMatch) {
    name = name.slice(0, duplicateMatch.index);
  }

  // Strip metadata suffix from the end.
  const metadataMatch = name.match(metadataSuffixPattern);
  if (metadataMatch) {
    name = name.slice(0, metadataMatch.index);
  }

  // Extract extension (e.g., .jpg from "photo.jpg").
  const extension = extname(name);
  if (extension) {
    name = basename(name, extension);
  }

  return { name, duplicate, extension: extension?.toLowerCase() };
}

File: packages/core/src/pixelkasten/stages/link.py
"""
Sidecar matching — link media files to their Google Takeout JSON metadata.

Ported from packages/core/src/stages/link.js. Handles Google's complex
filename truncation patterns documented in docs/takeout.md.

IMPORTANT: Uses os.path for filename decomposition, NOT pathlib. pathlib
silently normalizes double extensions (.MP.jpg) and duplicate markers (1).
"""

import os
import re

from pixelkasten.manifest import ManifestEntry, Sidecar, Source
from pixelkasten.configuration import Options

# Truncated variants of .supplemental-metadata (longest first for greedy matching).
_METADATA_SUFFIX_PATTERN = re.compile(
    "|".join(
        [
            r"\.supplemental-metadata",
            r"\.supplemental-metadat",
            r"\.supplemental-metada",
            r"\.supplemental-metad",
            r"\.supplemental-meta",
            r"\.supplemental-met",
            r"\.supplemental-me",
            r"\.supplemental-m",
            r"\.supplemental-",
            r"\.supplemental",
            r"\.supplementa",
            r"\.supplement",
            r"\.supplemen",
            r"\.suppleme",
            r"\.supplem",
            r"\.supple",
            r"\.suppl",
            r"\.supp",
            r"\.sup",
            r"\.su",
            r"\.s",
        ]
    )
    + "$"
)

# Truncated variants of -edited (longest first for greedy matching).
_EDITED_SUFFIX_PATTERN = re.compile(r"(?:-edited|-edite|-edit|-edi|-ed|-e)$")

# Duplicate marker: (N) at end of filename.
_DUPLICATE_PATTERN = re.compile(r"\((\d+)\)$")


def link(raw_collections: dict, options: Options) -> dict:
    """
    Match media files to their JSON sidecar metadata files.

    For each media file, attempts an exact match first (name + extension +
    duplicate marker), then falls back to fuzzy matching for truncated
    filenames if no exact match is found. In archive mode, sidecar matching
    is skipped entirely and every entry is loose.
    """
    files_media = raw_collections["files_media"]

    if options.mode == "archive":
        manifest = [
            ManifestEntry(media_path=fp, source=Source(type="loose"), sidecar=None)
            for fp in files_media
        ]
        return {
            "manifest": manifest,
            "stats": {
                "unmatched_metadata_files": set(),
                "unmatched_media_files": set(files_media),
            },
        }

    files_metadata = raw_collections["files_metadata"]
    files_metadata_albums = raw_collections.get("files_metadata_albums", [])

    # Stage 1: index metadata by directory for O(1) lookups.
    metadata_by_dir: dict[str, list[dict]] = {}
    for file_path in files_metadata:
        parsed = _parse_sidecar(file_path)
        dir_path = os.path.dirname(file_path)
        metadata_by_dir.setdefault(dir_path, []).append(parsed)

    # Stage 2: index albums by directory.
    albums: dict[str, str] = {}
    for path in files_metadata_albums:
        dir_path = os.path.dirname(path)
        albums[dir_path] = os.path.basename(dir_path)

    # Stage 3: pre-parse media files for performance.
    parsed_media = [_parse_media(fp) for fp in files_media]

    # Stage 4: match media files to their metadata sidecars.
    manifest: list[ManifestEntry] = []
    for media in parsed_media:
        dir_path = os.path.dirname(media["path"])
        candidates = metadata_by_dir.get(dir_path, [])

        source = (
            Source(type="album", name=albums[dir_path])
            if dir_path in albums
            else Source(type="loose")
        )

        sidecar_match = _match(media, candidates, options)
        sidecar = (
            Sidecar(path=sidecar_match["path"], confidence=sidecar_match["confidence"])
            if sidecar_match
            else None
        )

        manifest.append(
            ManifestEntry(
                media_path=media["path"],
                source=source,
                sidecar=sidecar,
            )
        )

    # Stage 5: compute statistics from the manifest.
    unmatched_metadata = set(files_metadata)
    unmatched_media = set(files_media)

    for entry in manifest:
        if entry.sidecar:
            unmatched_metadata.discard(entry.sidecar.path)
            unmatched_media.discard(entry.media_path)

    return {
        "manifest": manifest,
        "stats": {
            "unmatched_metadata_files": unmatched_metadata,
            "unmatched_media_files": unmatched_media,
        },
    }


def stripped_stem(file_path: str) -> str:
    """
    Return the media filename with extension, duplicate marker, and -edited
    suffix stripped. Used by link for matching and by group's archive-mode
    bucketing.
    """
    base = os.path.basename(file_path)
    name, _ = os.path.splitext(base)

    dup_match = _DUPLICATE_PATTERN.search(name)
    if dup_match:
        name = name[: dup_match.start()]

    edited_match = _EDITED_SUFFIX_PATTERN.search(name)
    if edited_match:
        name = name[: edited_match.start()]

    return name


def _parse_media(file_path: str) -> dict:
    """Parse a media file path into components for matching."""
    base = os.path.basename(file_path)
    _, extension = os.path.splitext(base)
    name, _ = os.path.splitext(base)

    dup_match = _DUPLICATE_PATTERN.search(name)
    duplicate = int(dup_match.group(1)) if dup_match else None

    return {
        "path": file_path,
        "name": stripped_stem(file_path),
        "duplicate": duplicate,
        "extension": extension.lower(),
    }


def _parse_sidecar(file_path: str) -> dict:
    """
    Parse a sidecar JSON file path into components for matching.

    Strips .json, duplicate marker, metadata suffix, then extracts the
    media extension from what remains.
    """
    base = os.path.basename(file_path)

    # Strip .json extension.
    name = os.path.splitext(base)[0]

    # Strip duplicate marker (N) from the end.
    dup_match = _DUPLICATE_PATTERN.search(name)
    duplicate = int(dup_match.group(1)) if dup_match else None
    if dup_match:
        name = name[: dup_match.start()]

    # Strip metadata suffix from the end.
    meta_match = _METADATA_SUFFIX_PATTERN.search(name)
    if meta_match:
        name = name[: meta_match.start()]

    # Extract extension (e.g., .jpg from "photo.jpg").
    name, extension = os.path.splitext(name)

    return {
        "path": file_path,
        "name": name,
        "duplicate": duplicate,
        "extension": extension.lower(),
    }


def _match(media: dict, candidates: list[dict], options: Options) -> dict | None:
    """Find the best metadata match for a media file within the same directory."""
    best_match = None
    best_score = 0

    for sidecar in candidates:
        # Strict duplicate check: (1) must always match (1).
        if media["duplicate"] != sidecar["duplicate"]:
            continue

        score = _match_score(media, sidecar, options)
        if score > best_score:
            best_score = score
            best_match = sidecar

    if not best_match:
        return None

    # Confidence levels based on match score.
    if best_score >= 150:
        confidence = 3
    elif best_score >= 100:
        confidence = 2
    else:
        confidence = 1

    return {"path": best_match["path"], "confidence": confidence}


def _match_score(media: dict, sidecar: dict, options: Options) -> int:
    """
    Calculate match score between a media file and a metadata file.

    Score hierarchy:
        150: Exact name + exact extension match (safest)
        100: Exact name match OR embedded extension match
        50+: Fuzzy/truncated name match (bidirectional prefix)
          0: No match
    """
    threshold = options.fuzzy_threshold
    fuzzy = options.fuzzy

    media_name = media["name"]
    media_ext = media["extension"]

    sidecar_name = sidecar["name"]
    sidecar_ext = sidecar["extension"]

    # Case 1: direct name match.
    if media_name == sidecar_name:
        if sidecar_ext and sidecar_ext == media_ext:
            return 150
        return 100

    # Case 2: embedded extension match. Handles cases like:
    #   media: "123.mp" (name="123", ext=".mp")
    #   metadata name: "123.mp"
    composite_name = f"{media_name}{media_ext}".lower()
    if composite_name == sidecar_name.lower():
        return 100

    # Fuzzy matching disabled.
    if not fuzzy:
        return 0

    # Case 3: fuzzy requires sufficient filename length.
    longer_name = max(len(media_name), len(sidecar_name))
    if longer_name < threshold:
        return 0

    # Case 4: bidirectional prefix match.
    sidecar_starts = sidecar_name.startswith(media_name)
    media_starts = media_name.startswith(sidecar_name)

    if media_starts or sidecar_starts:
        return 50 + min(len(media_name), len(sidecar_name))

    return 0
